Make llvm-nm's --export-symbols option work for ELF

bd1976bris opened the pull request:
The --export-symbols option..

> help string: _–export-symbols Export symbol list for all inputs_
> doc:  _Print sorted symbols with their visibility (if applicable), with duplicates removed._

.. was introduced in to llvm-nm for an AIX specific usecase. For non-XCOFF
target it currently simply lists all symbols. This seems incorrect given the option
name and intended functionality.

**ELF behaviour**
For ELF make this option display the set of exportable symbols. This is
particularly useful for bitcode as the binary tools do not offer an
easy way to dump the visibility of bitcode symbols. I have adjusted the
documentation to note that the linker ultimately decides which symbols
are exported and that --export-symbols can only show which symbols are
potential exports for ELF.

**SF_Exported for bitcode:**
I have modified the --export-symbols implementation to print only
symbols with SF_Exported. SF_Exported is currently set for object
file symbols but is not set for bitcode symbols. SF_Exported is checked
by some LLVM code and setting it for bitcode symbols might change
behaviour. So, instead of setting SF_Exported generally for bitcode 
symbols we instead set it for bitcode symbols in llvm-nm specific
code.

**Other file formats**
For non-XCOFF file formats I assume that --export-symbols is not
currently in use as it previously didn't produce sensible output.
Therefore, I do not have to worry about changing the behaviour for
those file formats. This change should improve the output for other
file formats and it should be easy to support them in future if
required. For now they remain unsupported and I have changed the
help text/documentation to note that.

bd1976bris committed:
Make llvm-nm's --export-symbols option work for ELF

The --export-symbols option..

–help:
  –export-symbols Export symbol list for all inputs\

doc:–export-symbols
  Print sorted symbols with their visibility (if applicable), with duplicates removed.

.. was introduced in to llvm-nm for an AIX specific usecase. For non-XCOFF target it simply
 lists all symbols which seems incorrect given the option name and intended functionality.

 This PR is for implementing the option for ELF targets so that it lists the set of unique
 exports. This is the set of non-local definitions that have `default` or `protected` ELF
 visibility.

MaskRay commented on llvm/tools/llvm-nm/llvm-nm.cpp:
```diff
@@ -777,7 +778,7 @@ static void printSymbolList(SymbolicFile &Obj,
 
   for (const NMSymbol &S : SymbolList) {
-    if (!S.shouldPrint())
+    if (!S.shouldPrint(false /*OnlyExported*/))
```
canonical style: `/*OnlyExported=*/false`

bd1976bris committed:
Implement --export-symbols for ELF

For ELF make this option display the set of exportable symbols. This is
particularly useful for bitcode as the binary tools do not offer an
easy way to dump the visibility of bitcode symbols. I have adjusted the
documentation to note that the linker ultimately decides which symbols
are exported and that --export-symbols can only show which symbols are
potential exports for ELF.

SF_Exported for bitcode:
SF_Exported is checked by some LLVM code and setting it for bitcode
symbols might change behaviour. Instead of setting SF_Exported
generally for bitcode symbols we instead set it for bitcode symbols
within llvm-nm.

For non-XCOFF file formats I assume that the --export-symbols is not in
not currently in use as it previously didn't produce sensible output.
Therefore, I do not have to worry about changing the behaviour for
those file formats. This change should improve the output for other
file formats and it should be easy to support them in future if
required. For now they are remain unsupported and I have changed the
help text/documentation to note that.

bd1976bris committed:
Minor improvements to wording
